homeassistant/components/tuya/entity.py

```diff
@@ -256,10 +256,40 @@ def get_dptype(
             order = ["function", "status_range"]
         for key in order:
             if dpcode in getattr(self.device, key):
-                return DPType(getattr(self.device, key)[dpcode].type)
+                try:
+                    return DPType(getattr(self.device, key)[dpcode].type)
+                except ValueError:
+                    fixed_type = TuyaEntity.determine_dptype(
+                        getattr(self.device, key)[dpcode].type
+                    )
+                    LOGGER.warning(
+                        f'Device {self.device.name} (id: {self.device.id}) has returned a bad model dpId type for code {dpcode} : "{getattr(self.device, key)[dpcode].type}", it should have been "{fixed_type}". Please contact the Tuya Support about this'
+                    )
+                    return fixed_type
 
         return None
 
+    @staticmethod
+    def determine_dptype(type) -> DPType | None:
+        """Determine the DPType.
+
+        Sometimes, we get ill formed DPTypes from the cloud,
+        this fixes them and maps them to the correct DPType
+        """
+        if type == "value":
+            return DPType(DPType.INTEGER)
+        if type in ("bitmap", "raw"):
+            return DPType(DPType.RAW)
+        if type == "enum":
+            return DPType(DPType.ENUM)
+        if type == "bool":
+            return DPType(DPType.BOOLEAN)
+        if type == "json":
+            return DPType(DPType.JSON)
+        if type == "string":
+            return DPType(DPType.STRING)
```

Comment:
instead of making this if/else tree, we can also make a lookup maybe?

Reply:
I'm not sure what you mean, do you mean a match/case?

Reply:
No, I meant making a dictionary to look them up directly.

Reply:
You can create a constant:
```python
_DPTYPE_MAPPING = {
    "bitmap": DPType.RAW,
    "raw": DPType.RAW,
    "value": DPType.INTEGER,
    ....
}
```

Then you can use the constant like this:
```python
for key in order:
    device_codes = getattr(self.device, key)
    if dpcode in device_codes:
        type = device_codes[dpcode].type
        try:
            return DPType(type)
        except ValueError:
            return _DPTYPE_MAPPING.get(type)
```

Reply:
Done, sorry for the back and forth :)

Reply:
No worries 👍